kable-core/src/androidMain/kotlin/BluetoothDeviceAndroidPeripheral.kt

```diff
@@ -68,198 +50,131 @@ internal class BluetoothDeviceAndroidPeripheral(
 
-    @Volatile
-    private var _connection: Connection? = null
-    private val connection: Connection
-        inline get() = _connection ?: throw NotReadyException(toString())
+    private val connection = MutableStateFlow<Connection?>(null)
+    private fun connectionOrThrow() =
+        connection.value
+            ?: throw NotConnectedException("Connection not established, current state: ${state.value}")
 
-    override val name: String? get() = bluetoothDevice.name
+    override val type: Type
+        get() = typeFrom(bluetoothDevice.type)
 
-    /**
-     * It's important that we instantiate this scope as late as possible, since [dispose] will be
-     * called immediately if the parent job is already complete. Doing so late in <init> is fine,
-     * but early in <init> it could reference non-nullable variables that are not yet set and crash.
-     */
-    private val scope = CoroutineScope(
-        parentCoroutineContext +
-            SupervisorJob(parentCoroutineContext.job).apply { invokeOnCompletion(::dispose) } +
-            CoroutineName("Kable/Peripheral/${bluetoothDevice.address}"),
-    )
+    override val address: String = bluetoothDevice.address
 
-    private val connectAction = scope.sharedRepeatableAction(::establishConnection)
+    @ExperimentalApi
+    override val name: String?
+        get() = bluetoothDevice.name
 
-    private suspend fun establishConnection(scope: CoroutineScope) {
-        checkBluetoothAdapterState(expected = STATE_ON)
-        bluetoothState.watchForDisablingIn(scope)
+    private suspend fun establishConnection(scope: CoroutineScope): CoroutineScope {
+        checkBluetoothIsOn()
 
         logger.info { message = "Connecting" }
         _state.value = State.Connecting.Bluetooth
 
         try {
-            _connection = bluetoothDevice.connect(
-                scope,
+            connection.value = bluetoothDevice.connect(
+                scope.coroutineContext,
                 applicationContext,
                 autoConnectPredicate(),
                 transport,
                 phy,
                 _state,
+                _services,
                 _mtu,
                 observers.characteristicChanges,
                 logging,
                 threadingStrategy,
+                disconnectTimeout,
             ) ?: throw ConnectionRejectedException()
 
-            suspendUntilOrThrow<State.Connecting.Services>()
+            suspendUntil<State.Connecting.Services>()
             discoverServices()
-            onServicesDiscovered(ServicesDiscoveredPeripheral(this@BluetoothDeviceAndroidPeripheral))
-
-            _state.value = State.Connecting.Observes
-            logger.verbose { message = "Configuring characteristic observations" }
-            observers.onConnected()
+            configureCharacteristicObservations()
         } catch (e: Exception) {
-            closeConnection()
```

Comment:
We no longer have to explicitly close the connection, because it will be closed implicitly when the connection coroutine scope tree completes. Since the connection process runs within a `Deferred` that is a child of the connection coroutine scope tree, any failure during connect will implicit close down the `Connection`.